src/peergos/shared/user/fs/CapabilityStore.java

```diff
@@ -280,115 +189,96 @@ private static boolean isEditCapFile(String filename) {
+                .thenCompose(seeked -> readSharingRecords(buf, ownerName, owner, seeked, file.getSize() - startOffset, network));
     }
 
-    private static CompletableFuture<Optional<CapabilityWithPath>> readSharingRecord(String ownerName,
-                                                                                     PublicKeyHash owner,
-                                                                                     AsyncReader reader,
-                                                                                     int offset,
-                                                                                     int capabilitySize,
-                                                                                     NetworkAccess network) {
-        byte[] serialisedFilePointer = new byte[capabilitySize];
-        return reader.seek( 0, offset).thenCompose( currentPos ->
-                currentPos.readIntoArray(serialisedFilePointer, 0, capabilitySize)
-                        .thenCompose(bytesRead -> {
-                            AbsoluteCapability pointer = AbsoluteCapability.fromCbor(CborObject.fromByteArray(serialisedFilePointer));
-                            EntryPoint entry = new EntryPoint(pointer, ownerName);
-                            return network.retrieveEntryPoint(entry).thenCompose( optFTN -> {
-                                if(optFTN.isPresent()) {
-                                    FileWrapper ftn = optFTN.get();
-                                    try {
-                                        return ftn.getPath(network)
-                                                .thenCompose(path -> CompletableFuture.completedFuture(Optional.of(new CapabilityWithPath(path, pointer))));
-                                    } catch (NoSuchElementException nsee) {
-                                        return Futures.errored(nsee); //file no longer exists
-                                    }
-                                } else {
-                                    return CompletableFuture.completedFuture(Optional.empty());
+    private static CompletableFuture<List<CapabilityWithPath>> readSharingRecords(byte[] buf,
+                                                                                  String ownerName,
+                                                                                  PublicKeyHash owner,
+                                                                                  AsyncReader reader,
+                                                                                  long maxBytesToRead,
+                                                                                  NetworkAccess network) {
+        if (maxBytesToRead == 0)
+            return CompletableFuture.completedFuture(Collections.emptyList());
+
+        List<AbsoluteCapability> caps = new ArrayList<>();
+        return reader.parseStream(AbsoluteCapability::fromCbor, caps::add, maxBytesToRead)
```

Comment:
should we version Capability?

Reply:
I think we're okay because all the components are themselves versioned, and any top level change we could add a version to the cbor map then, with an absent version being v0.